### PR TITLE
 Extending the OpenACC region in time integration split routine

### PR DESCRIPTION
**Pull request by brobey**
…on if block

 * Extending the OpenACC region to encompass all of the large iteration if block
 * Extending OpenACC region outside the if splitExplicitStep blocks
 * Forcing OpenACC loops to optimize
 * Fixing loop in diagnostic solve to run on GPU
 * Merging version that works with OpenACC and one that works with Intel
     - Intel fails on analysis test on Cori with the OpenACC version
     - OpenACC needs the array syntax replaced with traditional loops to
          optimize

[BFB]

### CODE REVIEW

**Comment by brobey** [comment]
Tests were run on Darwin with NVHPC 21.7, GCC 9, and Intel 19. Test was also run on the Cori KNL nodes with Intel. Tests passed bit for bit with previous version

**Comment by mark-petersen** [comment]
This compiles successfully with gnu and intel. Not surprising, as most lines are ACC pragmas, but still good to check.

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_split.F [hunk 1]
```diff
@@ -1787,22 +1800,62 @@ subroutine ocn_time_integrator_split(domain, dt)!{{{
             !$omp end parallel
 #endif
 
-            if (config_compute_active_tracer_budgets) then
+            if (varinp_compute_active_tracer_budgets) then
```
[the remark this answers is not in the document]
Comment (philipwjones) [reply components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_split.F:1803]: This is autogenerated (as with all the config variables). This is a temporary situation that was started when we replaced the scalar pointer retrievals with straight scalar retrievals, so these varinp versions are the scalar retrieved versions. We'll remove this duplication soon.

**Comment by brobey** [comment]
The varinp_compute_active_tracer_budgets variable is in a parallel region (kernels), so it must be a scalar and not a pointer. The duplication of the input variable is a temporary change to avoid breaking other packages.

--
Bob Robey
XCP-2, Computational Physics
505/412-8114 or 505/665-9052
Pronouns: he/him/his

From: Mark Petersen ***@***.***>
Reply-To: E3SM-Project/E3SM ***@***.***>
Date: Wednesday, March 30, 2022 at 6:19 AM
To: E3SM-Project/E3SM ***@***.***>
Cc: Robert Robey ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL] Re: [E3SM-Project/E3SM] Extending the OpenACC region in time integration split routine (PR #4853)


@mark-petersen commented on this pull request.

________________________________

In components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_split.F<https://urldefense.com/v3/__https:/github.com/E3SM-Project/E3SM/pull/4853*discussion_r838474289__;Iw!!Bt8fGhp8LhKGRg!XOlmWK6qerlPV9d2XZ53g6YoyA30cCgGtkdHCkvQi-B-QhUzR0Jgv4ecsPQT6rg$>:

> @@ -1787,22 +1800,62 @@ subroutine ocn_time_integrator_split(domain, dt)!{{{

             !$omp end parallel

 #endif



-            if (config_compute_active_tracer_budgets) then

+            if (varinp_compute_active_tracer_budgets) then

I don't see the variable varinp_compute_active_tracer_budgets defined anywhere. Is this a mistake?

**Review by mark-petersen — APPROVED** [review]
approved by inspection, and my own testing in stand-alone.

**Pull request closed without merging**